service-base/src/main/java/org/eclipse/hono/service/NewAbstractApplication.java

```diff
@@ -0,0 +1,276 @@
+
+        for (int i = 0; i < maxInstances; i++) {
+            final ConfigurationSupportingVerticle serviceInstance = serviceFactory.getObject();
+            log.debug("created new instance of service: " + serviceInstance);
+            final Future<String> tracker = Future.future();
+            vertx.deployVerticle(serviceInstance, tracker.completer());
+            deploymentTracker.add(tracker);
+        }
+
+        CompositeFuture.all(deploymentTracker).setHandler(s -> {
+            if (s.succeeded()) {
+                result.complete();
+            } else {
+                result.fail(s.cause());
+            }
+        });
+
+        return result;
+    }
+
+    /**
+     * Stops this application in a controlled fashion.
+     */
+    @PreDestroy
+    public final void shutdown() {
+//        final int shutdownTimeoutSeconds = config == null ? 20 : config.getStartupTimeout();
+        final int shutdownTimeoutSeconds = 20; // TODO: how to determine timeout? Really configurable per service?
+        shutdown(shutdownTimeoutSeconds, succeeded -> {
+            // do nothing
+        });
+    }
+
+    /**
+     * Stops this application in a controlled fashion.
+     *
+     * @param maxWaitTime The maximum time to wait for the server to shut down (in seconds).
+     * @param shutdownHandler The handler to invoke with the result of the shutdown attempt.
+     */
+    public final void shutdown(final long maxWaitTime, final Handler<Boolean> shutdownHandler) {
+
+        try {
+            preShutdown();
+            final CountDownLatch latch = new CountDownLatch(1);
+            if (vertx != null) {
+                log.debug("shutting down application...");
+                vertx.close(r -> {
+                    if (r.failed()) {
+                        log.error("could not shut down application cleanly", r.cause());
+                    }
+                    latch.countDown();
+                });
+            }
+            if (latch.await(maxWaitTime, TimeUnit.SECONDS)) {
+                log.info("application has been shut down successfully");
+                shutdownHandler.handle(Boolean.TRUE);
+            } else {
+                log.error("shut down timed out, aborting...");
+                shutdownHandler.handle(Boolean.FALSE);
+            }
+        } catch (InterruptedException e) {
+            log.error("shut down has been interrupted, aborting...");
+            Thread.currentThread().interrupt();
+            shutdownHandler.handle(Boolean.FALSE);
+        }
+    }
+
+    // TODO: How could we provide this?
+//    /**
+//     * Invoked with each created service instance.
+//     * <p>
+//     * Subclasses may override this method in order to customize
+//     * the service instance before it gets deployed to the Vert.x
+//     * container.
+//     *
+//     * @param instance The service instance.
+//     */
+//    protected void customizeServiceInstance(final S instance) {
```

Comment:
This can no longer be provided in an obvious way. What to do? Is it currently necessary, since it is not overridden anywhere?